job/server/src/main/java/alluxio/job/JobWorkerContext.java

```diff
@@ -20,23 +22,44 @@
  */
 @ThreadSafe
 public final class JobWorkerContext {
+  private final FileSystem mFileSystem;
+  private final FileSystemContext mFsContext;
   private final long mJobId;
   private final int mTaskId;
   private final UfsManager mUfsManager;
 
   /**
    * Creates a new instance of {@link JobWorkerContext}.
    *
+   * @param filesystem the filesystem client that is used when running jobs
+   * @param fsContext the filesystem client's underlying context
```

Comment:
This doesn't guarantee that the context is the file system's underlying context. But I don't think we can just pass in a context and create a filesystem because the caching does not work on that code path.

Reply:
I guess we _could_ only pass in the `FileSystemContext` instead  of both, but then we still have to call `FileSystem.Factory.create(FileSystemContext)`.

That isn't as bad as calling `FileSystem.Factory.create(AlluxioConfiguration)` because we aren't creating any new thread pools but it would be unnecessarily going through and iterating over all of our configuration objects (see [here](https://github.com/Alluxio/alluxio/blob/master/core/client/fs/src/main/java/alluxio/client/file/FileSystem.java#L134-L148) for what I mean).

Passing both is better performance-wise, but we technically don't have the guarantee that the fsContext belongs to the client.

Personally I really wish we had some kind of way to access the context from outside the client, but that probably requires re-architecting, and probably some kind of wrapper object